libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,522 @@
+import { RushConstants } from '../RushConstants';
+import { IOperationSettings, RushProjectConfiguration } from '../../api/RushProjectConfiguration';
+import { getHashesForGlobsAsync } from '../buildCache/getHashesForGlobsAsync';
+
+import type { Operation } from './Operation';
+import type { OperationExecutionManager } from './OperationExecutionManager';
+import type { OperationExecutionRecord } from './OperationExecutionRecord';
+import type {
+  IOperationRunnerAfterExecuteContext,
+  IOperationRunnerBeforeExecuteContext
+} from './OperationRunnerHooks';
+import type { IOperationRunner } from './IOperationRunner';
+import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import type {
+  ICreateOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import type { IRawRepoState, ProjectChangeAnalyzer } from '../ProjectChangeAnalyzer';
+import type { OperationMetadataManager } from './OperationMetadataManager';
+import type { BuildCacheConfiguration } from '../../api/BuildCacheConfiguration';
+import type { CobuildConfiguration } from '../../api/CobuildConfiguration';
+
+const PLUGIN_NAME: 'CacheablePhasedOperationPlugin' = 'CacheablePhasedOperationPlugin';
+
+export interface IOperationBuildCacheContext {
+  isCacheWriteAllowed: boolean;
+  isCacheReadAllowed: boolean;
+  isSkipAllowed: boolean;
+  projectBuildCache: ProjectBuildCache | undefined;
+  cobuildLock: CobuildLock | undefined;
+}
+
+export class CacheableOperationPlugin implements IPhasedCommandPlugin {
+  private _buildCacheContextByOperationRunner: Map<IOperationRunner, IOperationBuildCacheContext> = new Map<
+    IOperationRunner,
+    IOperationBuildCacheContext
+  >();
+
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tapPromise(
+      PLUGIN_NAME,
+      async (operations: Set<Operation>, context: ICreateOperationsContext): Promise<Set<Operation>> => {
+        const { buildCacheConfiguration, isIncrementalBuildAllowed } = context;
+        if (!buildCacheConfiguration) {
+          return operations;
+        }
+
+        for (const operation of operations) {
+          if (operation.runner) {
+            if (operation.runner instanceof ShellOperationRunner) {
+              const buildCacheContext: IOperationBuildCacheContext = {
+                // ShellOperationRunner supports cache writes by default.
+                isCacheWriteAllowed: true,
+                isCacheReadAllowed: isIncrementalBuildAllowed,
+                isSkipAllowed: isIncrementalBuildAllowed,
+                projectBuildCache: undefined,
+                cobuildLock: undefined
+              };
+              // Upstream runners may mutate the property of build cache context for downstream runners
+              this._buildCacheContextByOperationRunner.set(operation.runner, buildCacheContext);
```

Comment:
Only defining `_buildCacheContextByOperationRunner` if the runner is `ShellOperationRunner` is guaranteed to break caching in any repository where some operations have empty scripts, since empty scripts are cacheable but are handled by `NullOperationRunner` (to be the fast-path no-op).

The rushstack repo itself is a perfect example, since most packages depend on the rig packages, and the rig packages don't have build scripts.

Reply:
Nice catch!